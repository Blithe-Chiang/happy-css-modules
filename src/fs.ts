import * as fs from 'fs';


export function readFile(path: fs.PathLike): Promise<string> {
    return new Promise<string>((resolve, reject) => {
        fs.readFile(path, 'utf-8', (err, data) => {
            if (err !== null) {
                reject(err);
                return;
            }

            resolve(data);
        });
    });
}

export function writeFile(path: fs.PathLike, data: string): Promise<void> {
    return new Promise<void>((resolve, reject) => {
        fs.writeFile(path, data, 'utf-8', (err) => {
            if (err !== null) {
                reject(err);
                return;
            }

            resolve();
        });
    });
}
