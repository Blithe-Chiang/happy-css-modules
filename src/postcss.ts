import postcss, { Rule, AtRule, Root, Node } from 'postcss';
import modules from 'postcss-modules';
import selectorParser, { ClassName } from 'postcss-selector-parser';
import valueParser from 'postcss-value-parser';

/** The position of node. */
export type Position = {
  filePath: string;
  /** The line number in the source file. It is 1-based. */
  line?: number;
  /** The column number in the source file. It is 0-based. */
  column?: number;
};

/**
 * The name of exported token by the source file excluding the imported from other sheets.
 * @param ast The root node of the source file.
 * @returns The exported token names.
 */
export async function generateLocalTokenNames(ast: Root): Promise<string[]> {
  return new Promise((resolve, reject) => {
    postcss()
      .use(
        modules({
          getJSON: (_cssFileName, json) => {
            resolve(Object.keys(json));
          },
        }),
      )
      .process(ast)
      .catch(reject);
  });
}

/**
 * Generate the token's position on the source file.
 * @param rule The rule node that contains the token.
 * @param classSelector The class selector node that contains the token.
 * @returns The token's position on the source file.
 */
export function generateOriginalPosition(rule: Rule, classSelector: ClassName): Position {
  // The node derived from `postcss.parse` always has `source` property. Therefore, this line is unreachable.
  if (rule.source === undefined || classSelector.source === undefined) throw new Error('Node#source is undefined');
  // The node derived from `postcss.parse` always has `start` property. Therefore, this line is unreachable.
  if (rule.source.start === undefined || classSelector.source.start === undefined)
    throw new Error('Node#start is undefined');
  if (rule.source.input.file === undefined) throw new Error('Node#input.file is undefined');

  return {
    filePath: rule.source.input.file,
    // The line is 1-based.
    // TODO: If `source` contains an inline sourcemap, use the sourcemap to get the line and column.
    // This allows support for scss and less users.
    line: rule.source.start.line + classSelector.source.start.line - 1,
    // Postcss's column is 1-based but our column is 0-based.
    column: rule.source.start.column - 1 + (classSelector.source.start.column - 1),
  };
}

type Matcher = {
  atImport: (atImport: AtRule) => void;
  classSelector: (rule: Rule, classSelector: ClassName) => void;
};

function isAtRuleNode(node: Node): node is AtRule {
  return node.type === 'atrule';
}

function isAtImportNode(node: Node): node is AtRule {
  return isAtRuleNode(node) && node.name === 'import';
}

function isRuleNode(node: Node): node is Rule {
  return node.type === 'rule';
}

/**
 * Walk the AST and call the matcher functions.
 * @param ast The AST to walk.
 * @param matcher The matcher functions to call.
 */
export function walkByMatcher(ast: Root, matcher: Matcher): void {
  ast.walk((node) => {
    if (isAtImportNode(node)) {
      matcher.atImport(node);
    } else if (isRuleNode(node)) {
      // In `rule.selector` comes the following string:
      // 1. ".foo"
      // 2. ".foo:hover"
      // 3. ".foo, .bar"
      selectorParser((selectors) => {
        selectors.walk((selector) => {
          if (selector.type === 'class') {
            // In `selector.value` comes the following string:
            // 1. "foo"
            // 2. "bar"
            matcher.classSelector(node, selector);
          }
        });
      }).processSync(node);
    }
  });
}

/**
 * Parse the `@import` rule.
 * @param atImport The `@import` rule to parse.
 * @returns The imported sheet path.
 */
export function parseAtImport(atImport: AtRule): string | undefined {
  const firstNode = valueParser(atImport.params).nodes[0];
  if (firstNode.type === 'string') return firstNode.value;
  if (firstNode.type === 'function' && firstNode.value === 'url') {
    if (firstNode.nodes[0].type === 'string') return firstNode.nodes[0].value;
  }
}
