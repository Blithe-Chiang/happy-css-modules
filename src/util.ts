/**
 * The SystemError type of Node.js.
 * @see https://nodejs.org/api/errors.html#class-systemerror
 */
export interface SystemError {
  code: string;
}

export function isSystemError(value: unknown): value is SystemError {
  return (
    isObject(value) &&
    hasProp(value, 'constructor') &&
    isObject(value.constructor) &&
    hasProp(value.constructor, 'name') &&
    value.constructor.name === 'Error' &&
    hasProp(value, 'code') &&
    typeof value.code === 'string'
  );
}

export function isObject(value: unknown): value is object {
  return (typeof value === 'object' || typeof value === 'function') && value !== null;
}

export function hasProp<T extends string>(obj: object, prop: T): obj is { [key in T]: unknown } {
  return prop in obj;
}

export function unique<T>(array: T[]): T[] {
  return [...new Set(array)];
}

export function uniqueBy<T, U>(arr: T[], fn: (el: T) => U): T[] {
  const result: T[] = [];
  const keys = new Set<U>();
  for (const el of arr) {
    const key = fn(el);
    if (!keys.has(key)) {
      keys.add(key);
      result.push(el);
    }
  }
  return result;
}
