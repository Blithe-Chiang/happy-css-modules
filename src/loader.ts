import { readFile, stat } from 'fs/promises';
import { dirname, resolve } from 'path';
import postcss from 'postcss';
import {
  getOriginalLocation,
  generateLocalTokenNames,
  parseAtImport,
  Location,
  walkByMatcher,
  parseComposesDeclarationWithFromUrl,
} from './postcss';
import { unique } from './util';

/** The value returned from the transformer. */
export type TransformResult = {
  /** The transformed code. */
  css: string;
  /** The source map from the transformed code to the original code. */
  map?: string | object;
};

/** The function to transform source code. */
export type Transformer = (source: string, from: string) => TransformResult | Promise<TransformResult>;

/** The exported token. */
export type Token = {
  /** The token name. */
  name: string;
  /** The original locations of the token in the source file. */
  originalLocations: Location[];
};

type CacheEntry = {
  mtime: number;
  result: LoadResult;
};

type TokenImport =
  | {
      /** The `LoadResult` of the imported file. */
      fromResult: LoadResult;
      /** The import method. if `all`, import all tokens from the file. */
      type: 'all';
    }
  | {
      /** The `LoadResult` of the imported file. */
      fromResult: LoadResult;
      /** The import method. if `byNames`, import tokens from the file by the names. */
      type: 'byNames';
      /** The name of imported token. */
      names: string[];
    };

/** The result of `Loader#load`. */
export type LoadResult = {
  /** The file path of the source file. */
  filePath: string;
  /** The external tokens imported from the source file. */
  tokenImports: TokenImport[];
  /** The tokens exported by the source file. */
  localTokens: Token[];
};

/**
 * A class inspired by `css-modules-loader-core` to collect information about css modules.
 */
export class Loader {
  private cache: Map<string, CacheEntry> = new Map();
  async load(filePath: string, transform?: Transformer): Promise<LoadResult> {
    // If cache available, return it.
    // TODO: use `@file-cache/core`
    const mtime = (await stat(filePath)).mtime.getTime();
    const cacheEntry = this.cache.get(filePath);
    if (cacheEntry && cacheEntry.mtime === mtime) return cacheEntry.result;

    let css = await readFile(filePath, 'utf-8');
    let map: string | object | undefined;
    if (transform) {
      const result = await transform(css, filePath);
      css = result.css;
      map = result.map;
    }

    const ast = postcss.parse(css, { from: filePath, map: { inline: false, prev: map } });

    // Get the local tokens exported by the source file.
    // The tokens are fetched using `postcss-modules` plugin.
    const localTokenNames = await generateLocalTokenNames(ast);

    const importedSheetPaths: string[] = [];
    const filePathToImportedTokenNames = new Map<string, string[]>();
    const localTokens: Token[] = [];

    walkByMatcher(ast, {
      // Collect the sheets imported by `@import` rule.
      atImport: (atImport) => {
        const importedSheetPath = parseAtImport(atImport);
        if (importedSheetPath) importedSheetPaths.push(resolve(dirname(filePath), importedSheetPath));
      },
      // Traverse the source file to find a class selector that matches the local token.
      classSelector: (rule, classSelector) => {
        // Consider a class selector to be the origin of a token if it matches a token fetched by postcss-modules.
        // NOTE: This method has false positives. However, it works as expected in many cases.
        if (!localTokenNames.includes(classSelector.value)) return;

        const originalLocation = getOriginalLocation(rule, classSelector);

        const localToken = localTokens.find((token) => token.name === classSelector.value);
        if (localToken) {
          localToken.originalLocations.push(originalLocation);
        } else {
          localTokens.push({
            name: classSelector.value,
            originalLocations: [originalLocation],
          });
        }
      },
      composesDeclaration: (composesDeclaration) => {
        const result = parseComposesDeclarationWithFromUrl(composesDeclaration);
        if (result) {
          const from = resolve(dirname(filePath), result.from);
          const oldTokenNames = filePathToImportedTokenNames.get(from) ?? [];
          filePathToImportedTokenNames.set(from, [...oldTokenNames, ...result.tokenNames]);
        }
      },
    });

    const filePathToTokenImport = new Map<string, TokenImport>();

    // Load imported sheets recursively.
    for (const importedSheetPath of importedSheetPaths) {
      if (filePathToTokenImport.has(importedSheetPath)) continue;
      const fromResult = await this.load(importedSheetPath, transform);
      filePathToTokenImport.set(importedSheetPath, {
        fromResult,
        type: 'all',
      });
    }

    // Load imported tokens by the names.
    for (const [filePath, tokenNames] of filePathToImportedTokenNames) {
      const tokenImport = filePathToTokenImport.get(filePath);
      if (tokenImport) {
        if (tokenImport.type === 'all') {
          // noop
        } else {
          filePathToTokenImport.set(filePath, {
            fromResult: tokenImport.fromResult,
            type: 'byNames',
            names: unique([...tokenImport.names, ...tokenNames]),
          });
        }
      } else {
        filePathToTokenImport.set(filePath, {
          fromResult: await this.load(filePath, transform),
          type: 'byNames',
          names: unique(tokenNames),
        });
      }
    }

    const result: LoadResult = {
      filePath,
      tokenImports: [...filePathToTokenImport.values()],
      localTokens,
    };
    this.cache.set(filePath, { mtime, result });
    return result;
  }
}
