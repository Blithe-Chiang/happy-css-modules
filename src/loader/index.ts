import { readFile, stat } from 'fs/promises';
import { dirname, resolve } from 'path';
import postcss from 'postcss';
import { unique, uniqueBy } from '../util.js';
import {
  getOriginalLocation,
  generateLocalTokenNames,
  parseAtImport,
  type Location,
  parseComposesDeclarationWithFromUrl,
  collectNodes,
} from './postcss.js';

/**
 * The value returned from the transformer.
 * `false` means to skip transpiling on that file.
 * */
export type TransformResult =
  | {
      /** The transformed code. */
      css: string;
      /** The source map from the transformed code to the original code. */
      map: string | object;
      dependencies: (string | URL)[];
    }
  | false;

/** The function to transform source code. */
export type Transformer = (source: string, from: string) => TransformResult | Promise<TransformResult>;

/** The exported token. */
export type Token = {
  /** The token name. */
  name: string;
  /** The original locations of the token in the source file. */
  originalLocations: Location[];
};

type CacheEntry = {
  mtime: number; // TODO: `--cache-strategy` option will allow you to switch between `content` and `metadata` modes.
  result: LoadResult;
};

/** The result of `Loader#load`. */
export type LoadResult = {
  /** The path of the file imported from the source file with `@import` or `composes`. */
  dependencies: string[];
  /** The tokens exported by the source file. */
  tokens: Token[];
};

function normalizeTokens(tokens: Token[]): Token[] {
  const tokenNameToOriginalLocations = new Map<string, Location[]>();
  for (const token of tokens) {
    tokenNameToOriginalLocations.set(
      token.name,
      uniqueBy([...(tokenNameToOriginalLocations.get(token.name) ?? []), ...token.originalLocations], (location) =>
        JSON.stringify(location),
      ),
    );
  }
  return Array.from(tokenNameToOriginalLocations.entries()).map(([name, originalLocations]) => ({
    name,
    originalLocations,
  }));
}

/** This class collects information on tokens exported from CSS Modules files. */
export class Loader {
  private readonly cache: Map<string, CacheEntry> = new Map();
  private readonly transform: Transformer | undefined;

  constructor(transform?: Transformer) {
    // TODO: support resolver
    // TODO: support default transformer
    // TODO: support default resolver
    this.transform = transform;
  }

  /** Returns `true` if the cache is outdated. */
  private async isCacheOutdated(filePath: string): Promise<boolean> {
    const entry = this.cache.get(filePath);
    if (!entry) return true;
    const mtime = (await stat(filePath)).mtime.getTime();
    if (entry.mtime !== mtime) return true;

    const { dependencies } = entry.result;
    for (const dependency of dependencies) {
      const entry = this.cache.get(dependency);
      if (!entry) return true;
      const mtime = (await stat(dependency)).mtime.getTime();
      if (entry.mtime !== mtime) return true;
    }
    return false;
  }

  /**
   * Reads the source file and returns the code.
   * If transform is specified, the code is transformed before returning.
   */
  private async readCSS(
    filePath: string,
  ): Promise<
    | { css: string; map: undefined; dependencies: string[] }
    | { css: string; map: string | object | undefined; dependencies: string[] }
  > {
    const css = await readFile(filePath, 'utf-8');
    if (!this.transform) return { css, map: undefined, dependencies: [] };
    const result = await this.transform(css, filePath);
    if (result === false) return { css, map: undefined, dependencies: [] };
    return {
      css: result.css,
      map: result.map,
      dependencies: result.dependencies.map((dep) => {
        if (typeof dep === 'string') return dep;
        if (dep.protocol !== 'file:') throw new Error('Unsupported protocol: ' + dep.protocol);
        return dep.pathname;
      }),
    };
  }

  /** Returns information about the tokens exported from the CSS Modules file. */
  async load(filePath: string): Promise<LoadResult> {
    // NOTE: Loader does not support concurrent calls.
    // TODO: Throw an error if called concurrently.
    if (!(await this.isCacheOutdated(filePath))) {
      const cacheEntry = this.cache.get(filePath)!;
      return cacheEntry.result;
    }

    const mtime = (await stat(filePath)).mtime.getTime();

    const { css, map, dependencies } = await this.readCSS(filePath);

    const ast = postcss.parse(css, { from: filePath, map: { inline: false, prev: map } });

    // Get the local tokens exported by the source file.
    // The tokens are fetched using `postcss-modules` plugin.
    const localTokenNames = await generateLocalTokenNames(ast);

    const tokens: Token[] = [];

    const { atImports, classSelectors, composesDeclarations } = collectNodes(ast);

    // Load imported sheets recursively.
    for (const atImport of atImports) {
      const importedSheetPath = parseAtImport(atImport);
      if (!importedSheetPath) continue;
      const from = resolve(dirname(filePath), importedSheetPath);
      const result = await this.load(from);
      const externalTokens = result.tokens;
      dependencies.push(from);
      tokens.push(...externalTokens);
    }

    // Traverse the source file to find a class selector that matches the local token.
    for (const { rule, classSelector } of classSelectors) {
      // Consider a class selector to be the origin of a token if it matches a token fetched by postcss-modules.
      // NOTE: This method has false positives. However, it works as expected in many cases.
      if (!localTokenNames.includes(classSelector.value)) continue;

      const originalLocation = getOriginalLocation(rule, classSelector);

      tokens.push({
        name: classSelector.value,
        originalLocations: [originalLocation],
      });
    }

    // Load imported tokens by the names recursively.
    for (const composesDeclaration of composesDeclarations) {
      const declarationDetail = parseComposesDeclarationWithFromUrl(composesDeclaration);
      if (!declarationDetail) continue;
      const from = resolve(dirname(filePath), declarationDetail.from);
      const result = await this.load(from);
      const externalTokens = result.tokens.filter((token) => declarationDetail.tokenNames.includes(token.name));
      dependencies.push(from);
      tokens.push(...externalTokens);
    }

    const result: LoadResult = {
      dependencies: unique(dependencies).filter((dep) => dep !== filePath),
      tokens: normalizeTokens(tokens),
    };
    this.cache.set(filePath, { mtime, result });
    return result;
  }
}
