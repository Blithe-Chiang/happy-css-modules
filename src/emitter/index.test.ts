import { readFile, stat } from 'fs/promises';
import { EOL } from 'os';
import mock from 'mock-fs';
import { Token } from '../loader';
import { Location } from '../postcss';
import { exists } from '../test/util';
import {
  getDtsFilePath,
  getSourceMapFilePath,
  generateSourceMappingURLComment,
  generateDtsContentWithSourceMap,
  DtsFormatOptions,
  emitGeneratedFiles,
} from '.';

test('getDtsFilePath', () => {
  expect(getDtsFilePath('/app/src/dir/1.css', undefined)).toBe('/app/src/dir/1.css.d.ts');
  expect(getDtsFilePath('/app/src/dir/1.css', { rootDir: '/app', outDir: '/app/dist' })).toBe(
    '/app/dist/src/dir/1.css.d.ts',
  );
  expect(() => getDtsFilePath('/tmp/src/dir/1.css', { rootDir: '/app', outDir: '/app/dist' })).toThrow();
  expect(() => getDtsFilePath('/app/src/dir/1.css', { rootDir: '/app', outDir: '/tmp/dist' })).toThrow();
});

test('getSourceMapFilePath', () => {
  expect(getSourceMapFilePath('/app/src/dir/1.css', undefined)).toBe('/app/src/dir/1.css.d.ts.map');
  expect(getSourceMapFilePath('/app/src/dir/1.css', { rootDir: '/app', outDir: '/app/dist' })).toBe(
    '/app/dist/src/dir/1.css.d.ts.map',
  );
  expect(() => getSourceMapFilePath('/tmp/src/dir/1.css', { rootDir: '/app', outDir: '/app/dist' })).toThrow();
  expect(() => getSourceMapFilePath('/app/src/dir/1.css', { rootDir: '/app', outDir: '/tmp/dist' })).toThrow();
});

test('generateSourceMappingURLComment', () => {
  expect(generateSourceMappingURLComment('/app/src/dir/1.css.d.ts', '/app/src/dir/1.css.d.ts.map')).toBe(
    '//# sourceMappingURL=1.css.d.ts.map' + EOL,
  );
  expect(generateSourceMappingURLComment('/app/src/dir1/1.css.d.ts', '/app/src/dir2/1.css.d.ts.map')).toBe(
    '//# sourceMappingURL=../dir2/1.css.d.ts.map' + EOL,
  );
});

function fakeToken(args: {
  name: Token['name'];
  originalLocations: { filePath?: Location['filePath']; start: Location['start'] }[];
}): Token {
  return {
    name: args.name,
    originalLocations: args.originalLocations.map((location) => ({
      filePath: location.filePath ?? '/test/1.css',
      start: location.start,
      end: {
        line: location.start.line,
        column: location.start.column + args.name.length - 1,
      },
    })),
  };
}

describe('generateDtsContentWithSourceMap', () => {
  const filePath = '/test/1.css';
  const dtsFilePath = '/test/1.css.d.ts';
  const sourceMapFilePath = '/test/1.css.map';
  const dtsFormatOptions: DtsFormatOptions = {
    localsConvention: undefined,
    namedExport: false,
  };
  test('generate dts content with source map', () => {
    const tokens: Token[] = [
      fakeToken({ name: 'foo', originalLocations: [{ filePath: '/test/1.css', start: { line: 1, column: 1 } }] }),
      fakeToken({ name: 'bar', originalLocations: [{ filePath: '/test/1.css', start: { line: 2, column: 1 } }] }),

      fakeToken({
        name: 'baz',
        originalLocations: [
          { filePath: '/test/1.css', start: { line: 3, column: 1 } },
          { filePath: '/test/1.css', start: { line: 4, column: 1 } },
        ],
      }),
      fakeToken({ name: 'qux', originalLocations: [{ filePath: '/test/2.css', start: { line: 5, column: 1 } }] }),
      fakeToken({
        name: 'quux',
        originalLocations: [
          { filePath: '/test/2.css', start: { line: 6, column: 1 } },
          { filePath: '/test/2.css', start: { line: 7, column: 1 } },
        ],
      }),
    ];
    const { dtsContent, sourceMap } = generateDtsContentWithSourceMap(
      filePath,
      dtsFilePath,
      sourceMapFilePath,
      tokens,
      dtsFormatOptions,
    );
    expect(dtsContent).toMatchSnapshot();
    expect(sourceMap).toMatchSnapshot(); // TODO: Make snapshot human-readable
  });
  describe('format case', () => {
    const rawTokenList: Token[] = [
      fakeToken({ name: 'foo-bar', originalLocations: [{ start: { line: 1, column: 1 } }] }),
      fakeToken({ name: 'foo_bar', originalLocations: [{ start: { line: 2, column: 1 } }] }),
    ];
    test('undefined', () => {
      const { dtsContent } = generateDtsContentWithSourceMap(filePath, dtsFilePath, sourceMapFilePath, rawTokenList, {
        ...dtsFormatOptions,
        localsConvention: undefined,
      });
      expect(dtsContent).toMatchSnapshot();
    });
    test('camelCaseOnly', () => {
      const { dtsContent } = generateDtsContentWithSourceMap(filePath, dtsFilePath, sourceMapFilePath, rawTokenList, {
        ...dtsFormatOptions,
        localsConvention: 'camelCaseOnly',
      });
      expect(dtsContent).toMatchSnapshot();
    });
    test('camelCase', () => {
      const { dtsContent } = generateDtsContentWithSourceMap(filePath, dtsFilePath, sourceMapFilePath, rawTokenList, {
        ...dtsFormatOptions,
        localsConvention: 'camelCase',
      });
      expect(dtsContent).toMatchSnapshot();
    });
    test('dashesOnly', () => {
      const { dtsContent } = generateDtsContentWithSourceMap(filePath, dtsFilePath, sourceMapFilePath, rawTokenList, {
        ...dtsFormatOptions,
        localsConvention: 'dashesOnly',
      });
      expect(dtsContent).toMatchSnapshot();
    });
    test('dashes', () => {
      const { dtsContent } = generateDtsContentWithSourceMap(filePath, dtsFilePath, sourceMapFilePath, rawTokenList, {
        ...dtsFormatOptions,
        localsConvention: 'dashes',
      });
      expect(dtsContent).toMatchSnapshot();
    });
  });

  test('change export type by namedExport', () => {
    const tokens: Token[] = [
      fakeToken({ name: 'foo', originalLocations: [{ start: { line: 1, column: 1 } }] }),
      fakeToken({ name: 'bar', originalLocations: [{ start: { line: 2, column: 1 } }] }),
    ];
    const { dtsContent: dtsContentWithoutNamedExport, sourceMap: sourceMapWithoutNamedExport } =
      generateDtsContentWithSourceMap(filePath, dtsFilePath, sourceMapFilePath, tokens, {
        ...dtsFormatOptions,
        namedExport: false,
      });
    expect(dtsContentWithoutNamedExport).toMatchInlineSnapshot(`
      "declare const styles: {
        readonly \\"foo\\": string;
        readonly \\"bar\\": string;
      };
      export = styles;
      "
    `);
    expect(sourceMapWithoutNamedExport).toMatchSnapshot(); // TODO: Make snapshot human-readable
    const { dtsContent: dtsContentWithNamedExport, sourceMap: sourceMapWithNamedExport } =
      generateDtsContentWithSourceMap(filePath, dtsFilePath, sourceMapFilePath, tokens, {
        ...dtsFormatOptions,
        namedExport: true,
      });
    expect(dtsContentWithNamedExport).toMatchInlineSnapshot(`
      "export const __esModule: true;
      export const foo: string;
      export const bar: string;
      "
    `);
    expect(sourceMapWithNamedExport).toMatchSnapshot(); // TODO: Make snapshot human-readable
  });
  test('emit other directory', () => {
    const tokens: Token[] = [
      fakeToken({
        name: 'foo',
        originalLocations: [{ filePath: '/test/src/1.css', start: { line: 1, column: 1 } }],
      }),
      fakeToken({
        name: 'bar',
        originalLocations: [{ filePath: '/test/src/1.css', start: { line: 2, column: 1 } }],
      }),
    ];
    const { dtsContent, sourceMap } = generateDtsContentWithSourceMap(
      '/test/src/1.css',
      '/test/dist/1.css.d.ts',
      '/test/dist/1.css.d.ts.map',
      tokens,
      dtsFormatOptions,
    );
    expect(dtsContent).toMatchInlineSnapshot(`
      "declare const styles: {
        readonly \\"foo\\": string;
        readonly \\"bar\\": string;
      };
      export = styles;
      "
    `);
    expect(sourceMap).toMatchSnapshot(); // TODO: Make snapshot human-readable
  });
});

describe('emitGeneratedFiles', () => {
  const defaultArgs = {
    filePath: '/test/1.css',
    tokens: [fakeToken({ name: 'foo', originalLocations: [{ start: { line: 1, column: 1 } }] })],
    distOptions: undefined,
    emitDeclarationMap: true,
    dtsFormatOptions: undefined,
  };
  beforeEach(() => {
    mock({
      '/test': {}, // empty directory
    });
  });
  test('generates .d.ts and .d.ts.map', async () => {
    await emitGeneratedFiles({ ...defaultArgs });
    expect(await exists('/test/1.css.d.ts')).toBeTruthy();
    // A link to the source map is embedded.
    expect(await readFile('/test/1.css.d.ts', 'utf8')).toEqual(
      expect.stringContaining('//# sourceMappingURL=1.css.d.ts.map'),
    );
    expect(await exists('/test/1.css.d.ts.map')).toBeTruthy();
  });
  test('generates only .d.ts and .d.ts.map if emitDeclarationMap is false', async () => {
    await emitGeneratedFiles({ ...defaultArgs, emitDeclarationMap: false });
    expect(await exists('/test/1.css.d.ts')).toBeTruthy();
    // A link to the source map is not embedded.
    expect(await readFile('/test/1.css.d.ts', 'utf8')).toEqual(
      expect.not.stringContaining('//# sourceMappingURL=1.css.d.ts.map'),
    );
    expect(await exists('/test/1.css.d.ts.map')).toBeFalsy();
  });
  test('skips writing to disk if the generated files are the same', async () => {
    const tokens1 = [fakeToken({ name: 'foo', originalLocations: [{ start: { line: 1, column: 1 } }] })];
    await emitGeneratedFiles({ ...defaultArgs, tokens: tokens1 });
    const mtimeForDts1 = (await stat('/test/1.css.d.ts')).mtime;
    const mtimeForSourceMap1 = (await stat('/test/1.css.d.ts.map')).mtime;

    await emitGeneratedFiles({ ...defaultArgs, tokens: tokens1 });
    const mtimeForDts2 = (await stat('/test/1.css.d.ts')).mtime;
    const mtimeForSourceMap2 = (await stat('/test/1.css.d.ts.map')).mtime;
    expect(mtimeForDts1).toEqual(mtimeForDts2); // skipped
    expect(mtimeForSourceMap1).toEqual(mtimeForSourceMap2); // skipped

    const tokens2 = [fakeToken({ name: 'bar', originalLocations: [{ start: { line: 1, column: 1 } }] })];
    await emitGeneratedFiles({ ...defaultArgs, tokens: tokens2 });
    const mtimeForDts3 = (await stat('/test/1.css.d.ts')).mtime;
    const mtimeForSourceMap3 = (await stat('/test/1.css.d.ts.map')).mtime;
    expect(mtimeForDts1).not.toEqual(mtimeForDts3); // not skipped
    expect(mtimeForSourceMap1).not.toEqual(mtimeForSourceMap3); // not skipped
  });
});
