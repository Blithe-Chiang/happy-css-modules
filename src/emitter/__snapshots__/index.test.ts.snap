// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generateDtsContentWithSourceMap change export type by namedExport 2`] = `
Object {
  "file": "1.css.d.ts",
  "mappings": "AAAA;AAAA,E,SAAAA,K,SAAA;AAAA,E,SACAC,K,SADA;AAAA;AAAA",
  "names": Array [
    "foo",
    "bar",
  ],
  "sourceRoot": "",
  "sources": Array [
    "1.css",
  ],
  "version": 3,
}
`;

exports[`generateDtsContentWithSourceMap change export type by namedExport 4`] = `
Object {
  "file": "1.css.d.ts",
  "mappings": "AAAA;aAAAA,G,SAAA;aACAC,G,SADA",
  "names": Array [
    "foo",
    "bar",
  ],
  "sourceRoot": "",
  "sources": Array [
    "1.css",
  ],
  "version": 3,
}
`;

exports[`generateDtsContentWithSourceMap emit other directory 2`] = `
Object {
  "file": "1.css.d.ts",
  "mappings": "AAAA;AAAA,E,SAAAA,K,SAAA;AAAA,E,SACAC,K,SADA;AAAA;AAAA",
  "names": Array [
    "foo",
    "bar",
  ],
  "sourceRoot": "",
  "sources": Array [
    "../src/1.css",
  ],
  "version": 3,
}
`;

exports[`generateDtsContentWithSourceMap format case camelCase 1`] = `
"declare const styles: {
  readonly \\"foo-bar\\": string;
  readonly \\"fooBar\\": string;
  readonly \\"foo_bar\\": string;
  readonly \\"fooBar\\": string;
};
export = styles;
"
`;

exports[`generateDtsContentWithSourceMap format case camelCaseOnly 1`] = `
"declare const styles: {
  readonly \\"fooBar\\": string;
  readonly \\"fooBar\\": string;
};
export = styles;
"
`;

exports[`generateDtsContentWithSourceMap format case dashes 1`] = `
"declare const styles: {
  readonly \\"foo-bar\\": string;
  readonly \\"fooBar\\": string;
  readonly \\"foo_bar\\": string;
  readonly \\"foo_bar\\": string;
};
export = styles;
"
`;

exports[`generateDtsContentWithSourceMap format case dashesOnly 1`] = `
"declare const styles: {
  readonly \\"fooBar\\": string;
  readonly \\"foo_bar\\": string;
};
export = styles;
"
`;

exports[`generateDtsContentWithSourceMap format case undefined 1`] = `
"declare const styles: {
  readonly \\"foo-bar\\": string;
  readonly \\"foo_bar\\": string;
};
export = styles;
"
`;

exports[`generateDtsContentWithSourceMap generate dts content with source map 1`] = `
"declare const styles: {
  readonly \\"foo\\": string;
  readonly \\"bar\\": string;
  readonly \\"baz\\": string;
  readonly \\"baz\\": string;
  readonly \\"qux\\": string;
  readonly \\"quux\\": string;
  readonly \\"quux\\": string;
};
export = styles;
"
`;

exports[`generateDtsContentWithSourceMap generate dts content with source map 2`] = `
Object {
  "file": "1.css.d.ts",
  "mappings": "AAAA;AAAA,E,SAAAA,K,SAAA;AAAA,E,SACAC,K,SADA;AAAA,E,SAEAC,K,SAFA;AAAA,E,SAGAA,K,SAHA;AAAA,E,SCIAC,K,SDJA;AAAA,E,SCKAC,M,SDLA;AAAA,E,SCMAA,M,SDNA;AAAA;AAAA",
  "names": Array [
    "foo",
    "bar",
    "baz",
    "qux",
    "quux",
  ],
  "sourceRoot": "",
  "sources": Array [
    "1.css",
    "2.css",
  ],
  "version": 3,
}
`;
